proposals/lip-introduce_certificate_generation_mechanism.md

```diff
@@ -0,0 +1,843 @@
+
+Using the single commits collected by a node running blockchain *A*, the following function computes the certificate of largest height that can be submitted to another blockchain *B*.
+
+###### Parameters
+
+* `lastCertifiedHeight`: The height of the last certificate from blockchain *A* that was submitted to blockchain *B*.
+
+###### Returns
+
+The function returns a valid certificate object that can be submitted to blockchain *B* or `None` if no certificate of height larger than `lastCertifiedHeight` can be obtained from the single commits.
+
+###### Execution
+
+For readability, we describe the logic using two functions, the main function `getNextCertificateFromSingleCommits` and the auxiliary function `computeEligibleSingleCommits`.
+
+```python
+getNextCertificateFromSingleCommits(lastCertifiedHeight):
+    lastValidatorsHash  = validatorsHash property of block header at height lastCertifiedHeight
+    lastCertifiedValidators = validatorsHashPreimage[lastValidatorsHash].validators
+    lastCertificateThreshold = validatorsHashPreimage[lastValidatorsHash].certificateThreshold
+
+    addressToBFTWeight = {}
+    for validator in lastCertifiedValidators:
+        validatorAddress = validatorsModule.getAddressFromBLSKey(validator.blsKey)
+        addressToBFTWeight[validatorAddress] = validator.bftWeight
+
+    h = bftModule.getBFTHeights().maxHeightCertified
+    while h > lastCertifiedHeight:
+        if h in singleCommits:
+            # Obtain the subset of single commits from singleCommits[h] that are signed by a validator known to blockchain B.
+            # Note that the following function returns [] if the sum of BFT weights of these single commits does not reach
+            # the certificate threshold.
+            eligibleSingleCommits = computeEligibleSingleCommits(addressToBFTWeight, lastCertificateThreshold, singleCommits[h])
+            if eligibleSingleCommits != []:
+                aggregateCommit = aggregateSingleCommits(eligibleSingleCommits)
+                return getCertificateFromAggregateCommit(aggregateCommit)
+        h -= 1
+    return None
+```
+
+```python
+computeEligibleSingleCommits(addressToBFTWeight, lastCertificateThreshold, singleCommitsArray):
+    eligibleSingleCommits = []
+    aggregateBFTWeight = 0
+    for singleCommit in singleCommitsArray:
+        # certificate must only be signed by BLS keys known to the other chain
+        if singleCommit.validatorAddress in addressToBFTWeight:
+            append singleCommit to eligibleSingleCommits
+            aggregateBFTWeight += addressToBFTWeight[singleCommit.validatorAddress]
+    if aggregateBFTWeight >= lastCertificateThreshold:
+        return eligibleSingleCommits
+    else:
+        return []
+```
+
+
+[ccu-lip]: https://research.lisk.com/t/introduce-cross-chain-update-transactions/298
+[bft-module-lip]: https://research.lisk.com/
+[block-header-lip]:  https://research.lisk.com/t/update-block-schema-and-block-processing/293
```

Comment:
https://github.com/LiskHQ/lips/blob/main/proposals/lip-0055.md

Reply:
In general many of these links could already be updated

Reply:
Changed.